3rd-party/cove/app/mainform.cpp

```diff
@@ -892,6 +842,31 @@ void mainForm::on_applyFIRFilterPushButton_clicked()
 	if (!newImageBitmap.isNull()) imageBitmap = newImageBitmap;
 	ui.imageView->setImage(&imageBitmap);
 }
+
+/* \class mainForm::BwBitmapUndoStep
+ * \brief Embedded struct holding undo data.
+ */
+
+mainForm::BwBitmapUndoStep::BwBitmapUndoStep(mainForm* form, QImage image, bool vectorizable)
+    : formPtr { form }
+    , image { image }
+    , suitableForVectorization { vectorizable }
+{
+	// nothing
+}
+
+void mainForm::BwBitmapUndoStep::redo()
+{
+	undo(); // the same data swap
+}
+
+void mainForm::BwBitmapUndoStep::undo()
+{
+	Q_ASSERT(formPtr);
+	std::swap(formPtr->bwBitmap, image);
+	std::swap(formPtr->bwBitmapVectorizable, suitableForVectorization);
```

Comment:
Note that idiomatic use of swap is:
~~~
using std::swap;
swap(formPtr->bwBitmap, image);
swap(formPtr->bwBitmapVectorizable, suitableForVectorization);
~~~
in order to find optimized specializations via [ADL](https://en.cppreference.com/w/cpp/language/adl). However,  QImage doesn't provide such a specialization (free function). It has a member function for swap.

Reply:
Understood. I didn't know this idiom. Thanks!